### PR TITLE
Implement external flags & Local defaults

### LINKED ISSUE #864: implement --rpc-external and --ws-external flags
## Expected Behavior

- rpc/ws servers are open to local connections by default, and can be opened to external connections using `--rpc-external` and `--ws-external`

## Current Behavior

- rpc/ws servers are off by default (I think this is ok), can be turned on using the `--rpc` and `--ws` flags.

## Possible Solution

- change rpc/ws server to accept only local connections unless specified by the given external flag.

### PR DESCRIPTION
**Pull request by RyRy79261**
## Changes

- `--rpc` flag changed to `--rpc-external`
- `--ws` flag changed to `--ws-external`

## Tests

```

```

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [x] I have reviewed my own pull request before requesting a review
- [ ] All integration tests and required coverage checks are passing

## Issues

closes #864

### CODE REVIEW

File: dot/rpc/http.go [hunk 1]
```diff
@@ -138,15 +138,13 @@ func (h *HTTPServer) Start() error {
...
-	if !h.serverConfig.WSEnabled {
-		return nil
-	}
+	// TODO: Confirm if RPC & WS always enabled is correct
 
 	h.logger.Info("Starting WebSocket Server...", "host", h.serverConfig.Host, "port", h.serverConfig.WSPort)
 	ws := mux.NewRouter()
 	ws.Handle("/", h)
 	go func() {
+		// TODO: Confirm if this only accepts localhost
```
Comment (RyRy79261) [inline dot/rpc/http.go:147]: @noot

File: dot/rpc/http.go [hunk 2]
```diff
@@ -138,15 +138,13 @@ func (h *HTTPServer) Start() error {
 			h.logger.Error("http error", "err", err)
 		}
 	}()
-
-	if !h.serverConfig.WSEnabled {
-		return nil
-	}
+	// TODO: Confirm if RPC & WS always enabled is correct
```
Comment (RyRy79261) [inline dot/rpc/http.go:141]: @noot

File: cmd/gossamer/flags.go [hunk 3]
```diff
@@ -155,14 +160,21 @@ var (
...
 	}
+	// WSEnabledFlag Enable the websockets server
 	WSEnabledFlag = cli.BoolFlag{
-		Name:  "ws",
+		Name:  "ws-enabled",
 		Usage: "Enable the websockets server",
 	}
+	// WSExternalEnabledFlag Enable external websocket connections
+	WSExternalEnabledFlag = cli.BoolFlag{
+		Name:  "ws-external-enabled",
```
Comment (noot) [inline cmd/gossamer/flags.go:175]: can you rename this to `--ws-external`?

File: dot/config.go [hunk 4]
```diff
@@ -120,6 +122,21 @@ func RPCServiceEnabled(cfg *Config) bool {
...
+// WSServiceEnabled returns true if the websocket service is enabled
+func WSServiceEnabled(cfg *Config) bool {
+	return cfg.RPC.WSEnabled
+}
+
+// WSExternalServiceEnabled returns true if the websocket service accepts external connections
+func WSExternalServiceEnabled(cfg *Config) bool {
+	return cfg.RPC.WSExternalEnabled
+}
+
```
Comment (noot) [inline dot/config.go:139]: why is this needed? you should be able to directly the fields from the `Config` since they're exported

File: dot/rpc/http.go (hunk 1 above)
Comment (noot) [reply dot/rpc/http.go:153]: yeah don't really know if this is testable with unit tests... if you replaced 127.0.0.1 with your external IP it would probably resolve to localhost anyways right?

File: dot/rpc/websocket.go [hunk 5]
```diff
@@ -76,9 +77,19 @@ var rpcHost string
 
 // ServeHTTP implemented to handle WebSocket connections
 func (h *HTTPServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	fmt.Println("Entered WS Serve")
```
Comment (noot) [inline dot/rpc/websocket.go:80]: pls remove

File: dot/rpc/websocket.go [hunk 6]
```diff
@@ -76,9 +77,19 @@ var rpcHost string
...
 	var upg = websocket.Upgrader{
 		CheckOrigin: func(r *http.Request) bool {
-			return true // todo determine how this should check orgigin
+			// TODO: #864
+			if !h.serverConfig.WSExternalEnabled {
+				ip := GetIP(r)
+				fmt.Println("Requesting IP", ip)
```
Comment (noot) [inline dot/rpc/websocket.go:86]: same

File: dot/services.go [hunk 7]
```diff
@@ -274,7 +274,7 @@ func createRPCService(cfg *Config, stateSrvc *state.Service, coreSrvc *core.Serv
 		"host", cfg.RPC.Host,
 		"rpc port", cfg.RPC.Port,
 		"mods", cfg.RPC.Modules,
-		"ws enabled", cfg.RPC.WSEnabled,
```
Comment (noot) [inline dot/services.go:277]: can you leave this line in?

File: scripts/integration-test-all.sh [hunk 8]
```diff
@@ -57,7 +57,7 @@ arr=()
 start_func() {
   echo "starting gossamer node $i in background ..."
   "$PWD"/bin/gossamer --port=$(($PORT + $i)) --key=${keys[$i-1]} --basepath="$BASE_PATH$i" \
-    --rpc --rpchost=$HOSTNAME --rpcport=$(($RPC_PORT + $i)) --roles=1 --rpcmods=system,author,chain >"$BASE_PATH"/node"$i".log 2>&1 & disown
+    --rpc --ws --rpchost=$HOSTNAME --rpcport=$(($RPC_PORT + $i)) --roles=1 --rpcmods=system,author,chain >"$BASE_PATH"/node"$i".log 2>&1 & disown
```
Comment (noot) [inline scripts/integration-test-all.sh:60]: why is this updated?

File: tests/utils/gossamer_utils.go [hunk 9]
```diff
@@ -447,10 +448,12 @@ func generateDefaultConfig() *ctoml.Config {
...
-			Enabled:   false,
-			Host:      "localhost",
-			Modules:   []string{"system", "author", "chain", "state"},
-			WSEnabled: false,
+			Enabled:           false,
+			ExternalEnabled:   false,
+			Host:              "localhost",
+			Modules:           []string{"system", "author", "chain", "state"},
+			WSEnabled:         false,
+			WSExternalEnabled: false,
```
Comment (noot) [inline tests/utils/gossamer_utils.go:456]: same here, this can stay as before, the integration tests don't use ws at the moment

File: cmd/gossamer/flags.go [hunk 10]
```diff
@@ -155,14 +160,21 @@ var (
...
+	// WSPortFlag WebSocket server listening port
 	WSPortFlag = cli.IntFlag{
 		Name:  "wsport",
 		Usage: "Websockets server listening port",
 	}
+	// WSEnabledFlag Enable the websockets server
 	WSEnabledFlag = cli.BoolFlag{
-		Name:  "ws",
+		Name:  "ws-enabled",
 		Usage: "Enable the websockets server",
```
Comment (noot) [inline cmd/gossamer/flags.go:171]: can you leave this as `ws`?

File: dot/rpc/http.go (hunk 1 above)
Comment (RyRy79261) [reply dot/rpc/http.go:153]: @noot I think I'd probably need to decouple the validation logic and then run tests against that rather than the logic that branches from it

File: dot/config.go (hunk 4 above)
Comment (RyRy79261) [reply dot/config.go:139]: @noot Added them there purely because I saw other config items exported in a similar way, will remove

File: dot/rpc/websocket.go (hunk 5 above)
Comment (RyRy79261) [reply dot/rpc/websocket.go]: Still working on the Regex stuff

File: scripts/integration-test-all.sh (hunk 8 above)
Comment (RyRy79261) [reply scripts/integration-test-all.sh:60]: With the services being disabled by default, I assumed that locally running would require both

File: dot/rpc/http.go [hunk 11]
```diff
@@ -132,21 +138,21 @@ func (h *HTTPServer) Start() error {
...
 	go func() {
 		err := http.ListenAndServe(fmt.Sprintf(":%d", h.serverConfig.RPCPort), r)
 		if err != nil {
 			h.logger.Error("http error", "err", err)
 		}
 	}()
 
-	if !h.serverConfig.WSEnabled {
-		return nil
-	}
```
Comment (noot) [inline dot/rpc/http.go:144]: can you re-add this? otherwise ws will turn on whenever `--rpc` is set

File: dot/rpc/http.go [hunk 12]
```diff
@@ -132,21 +138,21 @@ func (h *HTTPServer) Start() error {
...
-
 	h.logger.Info("Starting WebSocket Server...", "host", h.serverConfig.Host, "port", h.serverConfig.WSPort)
 	ws := mux.NewRouter()
 	ws.Handle("/", h)
 	go func() {
+		if h.serverConfig.WSExternal {
+
+		}
```
Comment (noot) [inline dot/rpc/http.go:155]: does the validate request func from above work here or no?

Comment (noot) [reply dot/rpc/http.go:155]: ah I see you added it in `ServeHTTP`, can you remove this then?

File: cmd/gossamer/config.go [hunk 13]
```diff
@@ -553,6 +555,13 @@ func setDotRPCConfig(ctx *cli.Context, tomlCfg ctoml.RPCConfig, cfg *dot.RPCConf
 		cfg.Enabled = false
 	}
 
+	// check --rpc-external flag and update node configuration
+	if external := ctx.GlobalBool(RPCExternalFlag.Name); external {
+		cfg.External = true
```
Comment (noot) [inline cmd/gossamer/config.go:560]: could you update this to also set `cfg.Enabled = true`? ie. if someone specifies `--rpc-external` but not `--rpc` it should still turn on the server

File: cmd/gossamer/config.go [hunk 14]
```diff
@@ -572,10 +581,16 @@ func setDotRPCConfig(ctx *cli.Context, tomlCfg ctoml.RPCConfig, cfg *dot.RPCConf
...
-	} else if ctx.IsSet(WSEnabledFlag.Name) && !wsenabled {
-		cfg.WSEnabled = false
+	if WS := ctx.GlobalBool(WSFlag.Name); WS {
+		cfg.WS = true
+	} else if ctx.IsSet(WSFlag.Name) && !WS {
+		cfg.WS = false
+	}
+
+	if wsExternal := ctx.GlobalBool(WSExternalFlag.Name); wsExternal {
+		cfg.WSExternal = true
```
Comment (noot) [inline cmd/gossamer/config.go:591]: same here, could you also add `cfg.WS = true`?

**Review by noot — COMMENTED** [review]
overall looking good! just a couple small changes

File: dot/rpc/helpers.go [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+		BlockByDefault: true,
+		AllowedIPs:     []string{"127.0.0.1", "::1"},
+	})
+}
+
+// LocalRequestOnly HTTP handler to restrict to only local connections
+func LocalRequestOnly(r *rpc.RequestInfo, i interface{}) error {
+	ip, _, error := net.SplitHostPort(r.Request.RemoteAddr)
+	if error != nil {
+		logger.Error("unable to parse IP", "error")
```
Comment (noot) [inline dot/rpc/helpers.go:39]: you can remove the log since the error is returned

File: dot/rpc/helpers.go [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+		return errors.New("unable to parse IP")
+	}
+	f := LocalhostFilter()
+	if allowed := f.Allowed(ip); allowed {
+		return nil
+	}
+	logger.Error("external HTTP request refuesed", "error")
```
Comment (noot) [inline dot/rpc/helpers.go:46]: same here

File: dot/rpc/websocket.go [hunk 17]
```diff
@@ -78,7 +80,22 @@ var rpcHost string
...
+					logger.Error("unable to parse IP", "error")
+					return false
+				}
+
+				f := LocalhostFilter()
+				if allowed := f.Allowed(ip); allowed {
+					return true
+				}
+
+				logger.Error("external websocket request refused", "error")
```
Comment (noot) [inline dot/rpc/websocket.go:95]: you could change this to debug since its not really an error

**Review by noot — APPROVED** [review]
looks good! please update logs

**Review by arijitAD — APPROVED** [review]
LGTM

File: dot/rpc/http_test.go [hunk 18]
```diff
@@ -90,3 +90,45 @@ func TestNewHTTPServer(t *testing.T) {
...
+	require.Nil(t, err)
+
+	req.Header.Set("Content-Type", "application/json")
+
+	res, err := client.Do(req)
+	require.Nil(t, err)
+	defer res.Body.Close()
+
+	require.Equal(t, "200 OK", res.Status)
+
```
Comment (edwardmack) [inline dot/rpc/http_test.go:133]: This test is called TestBlockExternalIP, but it doesn't seem to be testing an external IP. It does seem to confirm a local connection works. I see `server.rpcServer.RegisterValidateRequestFunc(LocalRequestOnly)` which seem like it's doing what it should, but I don't see how this is tested.

**Review by edwardmack — COMMENTED** [review]
Is there a way that we can test to confirm that external IP's are blocked?  I don't know how we'd setup a test for that, but it might be useful.

File: dot/rpc/http_test.go (hunk 18 above)
Comment (RyRy79261) [reply dot/rpc/http_test.go:133]: Ahh good spot, this was my initial set up test but wasn't sure how to simulate external IP addresses, removing for now

**Comment by RyRy79261** [comment]
@noot 

Something odd is happening with the Stress tests, its claiming that the newly added toml entries are missing despite the entries being present, not sure if theres a different source for these test's configs
image

**Comment by noot** [comment]
@RyRy79261  in `tests/utils/gossamer_utils.go`:
```
func generateDefaultConfig() *ctoml.Config {
	return &ctoml.Config{
		Global: ctoml.GlobalConfig{
			Name:   "gssmr",
			ID:     "gssmr",
			LogLvl: "crit",
		},
		Log: ctoml.LogConfig{
			CoreLvl: "info",
			SyncLvl: "info",
		},
		Init: ctoml.InitConfig{
			GenesisRaw: "./chain/gssmr/genesis-raw.json",
		},
		Account: ctoml.AccountConfig{
			Key:    "",
			Unlock: "",
		},
		Core: ctoml.CoreConfig{
			Roles:            4,
			BabeAuthority:    true,
			GrandpaAuthority: true,
		},
		Network: ctoml.NetworkConfig{
			Bootnodes:   nil,
			ProtocolID:  "/gossamer/gssmr/0",
			NoBootstrap: false,
			NoMDNS:      false,
		},
		RPC: ctoml.RPCConfig{
			Enabled:   false,
			Host:      "localhost",
			Modules:   []string{"system", "author", "chain", "state"},
			WSEnabled: false,
		},
	}
}
```

try changing `WSEnabled` to `WS`?

**Comment by RyRy79261** [comment]
@noot 

This is what I'm seeing on my machine: 
image

**Comment by noot** [comment]
try removing that field since it's false anyways

**Comment by RyRy79261** [comment]
Still receiving the same error from `make it-stress` 
image

**Comment by noot** [comment]
@RyRy79261 where is that error from? I just ran it locally and the test fails for other reasons

**Comment by RyRy79261** [comment]
@noot This is when I run the tests locally

**Comment by noot** [comment]
@RyRy79261 you see it in the output of `make it-stress`? it looks like it's from the log files

what's your local `chain/gssmr/config.toml`?

**Comment by RyRy79261** [comment]
@noot Yeah thats in my terminal after running the `make it-stress` command 

heres the config:
```[global]
name = "gssmr"
id = "gssmr"
basepath = "~/.gossamer/gssmr"
log = "info"

[log]
core = ""
network = ""
rpc = ""
state = ""
runtime = ""
babe = ""
grandpa = ""

[init]
genesis-raw = "./chain/gssmr/genesis-raw.json"

[account]
key = ""
unlock = ""

[core]
roles = 4
babe-authority = true
grandpa-authority = true
babe-threshold = ""

[network]
port = 7001
bootnodes = []
protocol = "/gossamer/gssmr/0"
nobootstrap = false
nomdns = false

[rpc]
enabled = false
external = false
port = 8545
host = "localhost"
modules = ["system", "author", "chain", "state"]
ws-port = 8546
ws = false
ws-external = false
```

**Comment by noot** [comment]
can you try removing the `ws = false` from there?

**Comment by RyRy79261** [comment]
@noot it throws on the external option now:
image

**Comment by noot** [comment]
@RyRy79261 okay try removing that also? it seems like something is off though, I didn't see those errors

**Comment by RyRy79261** [comment]
@noot Running into a opaque issue about not being able to start a node

I've restarted my machine twice 
image

**Comment by noot** [comment]
look at the logs in /tmp/, I saw the same issue on my side
